crates/bevy_ecs/src/schedule/stage.rs

```diff
@@ -199,51 +227,102 @@ impl SystemStage {
         &self.exclusive_before_commands
     }
 
-    // TODO: consider exposing
-    fn add_system_to_set(&mut self, system: impl Into<SystemDescriptor>, set: usize) -> &mut Self {
-        // This assertion is there to document that a maximum of `u32::MAX / 8` systems should be
-        // added to a stage to guarantee that change detection has no false positive, but it
-        // can be circumvented using exclusive or chained systems
-        assert!(
-            self.exclusive_at_start.len()
-                + self.exclusive_before_commands.len()
-                + self.exclusive_at_end.len()
-                + self.parallel.len()
-                < (u32::MAX / 8) as usize
-        );
+    pub fn with_system_set(mut self, system_set: SystemSet) -> Self {
+        self.add_system_set(system_set);
+        self
+    }
+
+    pub fn add_system_set(&mut self, system_set: SystemSet) -> &mut Self {
         self.systems_modified = true;
-        match system.into() {
-            SystemDescriptor::Exclusive(descriptor) => {
-                let insertion_point = descriptor.insertion_point;
-                let container = ExclusiveSystemContainer::from_descriptor(descriptor, set);
-                match insertion_point {
-                    InsertionPoint::AtStart => {
-                        let index = self.exclusive_at_start.len();
-                        self.uninitialized_at_start.push(index);
-                        self.exclusive_at_start.push(container);
-                    }
-                    InsertionPoint::BeforeCommands => {
-                        let index = self.exclusive_before_commands.len();
-                        self.uninitialized_before_commands.push(index);
-                        self.exclusive_before_commands.push(container);
-                    }
-                    InsertionPoint::AtEnd => {
-                        let index = self.exclusive_at_end.len();
-                        self.uninitialized_at_end.push(index);
-                        self.exclusive_at_end.push(container);
-                    }
-                }
-            }
-            SystemDescriptor::Parallel(descriptor) => {
-                self.uninitialized_parallel.push(self.parallel.len());
-                self.parallel
-                    .push(ParallelSystemContainer::from_descriptor(descriptor, set));
-            }
+        let (run_criteria, mut systems) = system_set.bake();
+        let set_run_criteria_label =
+            run_criteria.map(|criteria| self.process_run_criteria(criteria));
+        for system in systems.drain(..) {
+            self.add_system_inner(system, set_run_criteria_label.clone());
         }
         self
     }
 
+    pub fn with_stage_run_criteria<S: System<In = (), Out = ShouldRun>>(
```

Comment:
I'm not a huge fan of this name change. Ideally the "stage" context is implicit, just like it is for Systems and SystemSets (ex: we don't say `SystemSet::with_system_set_run_criteria`). This feels inconsistent in that regard.

I get that RunCriteria are now owned by Stages, but I expect that manually adding criteria directly to stages will be relatively uncommon. Maybe its better to keep `with_run_criteria` as "Stage criteria" and then require `Stage::add_system_run_criteria()` for dangling system criteria?

Reply:
> require Stage::add_system_run_criteria() for dangling system criteria?

This is a mouthful for fairly useful behavior. Ideally this would be a method on the system itself (or a label once we have "label properties"), but I suspect this isn't technically feasible right now.

Reply:
>  Ideally this would be a method on the system itself (or a label once we have "label properties"), but I suspect this isn't technically feasible right now.

This is already a method on the system itself : `foo.system().with_run_criteria(X)`.

The "problem" being solved by the stage method equivalent is that Systems no longer own their run criteria. They only "add" criteria to the Stage (which is then deduped when there are multiple identical labels) or reference criteria that already exists. The stage method allows users to create the run criteria directly in the Stage (the real owning container), then reference it via labels in systems. This is clearer than adding a "shared" criteria in the first system with a label, then _just_ adding the label in the second system.

Reply:
Ah, I see. In that case, I don't mind the longer name at all.

Reply:
I used the verb "define" for the "add a new criteria descriptor to stage" method at some point, I think I couldn't figure out how to name the `with` and `add` versions of it.

Reply:
Renamed back in c5e8f86d350e781c433eb4f556c17f0e06bb1f22